{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import sys\n",
    "import hcrfish"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example: Drosophila Melanogaster"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dm6.fa.gz\n",
      "    45153922 100%  559.25MB/s    0:00:00 (xfer#1, to-check=0/1)\n",
      "\n",
      "sent 47134 bytes  received 27010 bytes  29657.60 bytes/sec\n",
      "total size is 45153922  speedup is 609.00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gunzip: raw_data/dmel/genome/dm6.fa already exists -- skipping\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='chflags nohidden raw_data/dmel/genome/dm6.fa raw_data/dmel/genome/dm6.fa', returncode=0)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make directory \n",
    "output_dir = \"input/dmel/genome\"\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "# Download the genome\n",
    "rsync_path = \"rsync://hgdownload.soe.ucsc.edu/goldenPath/dm6/bigZips/dm6.fa.gz\"\n",
    "output_path = os.path.join(output_dir, \"dm6.fa.gz\")\n",
    "command = f\"rsync --partial --progress {rsync_path} {output_path}\"\n",
    "subprocess.run(command, shell=True)\n",
    "\n",
    "# Unzip the genome\n",
    "output_path = \"input/dmel/genome/dm6.fa.gz\"\n",
    "command = f\"gunzip {output_path}\"\n",
    "subprocess.run(command, shell=True)\n",
    "\n",
    "# Unhide genome file \n",
    "output_path = \"input/dmel/genome/dm6.fa\"\n",
    "command = f\"chflags nohidden {output_path} {output_path}\" \n",
    "subprocess.run(command, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download Transcriptome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dm6.ncbiRefSeq.gtf.gz\n",
      "     5133989 100%  816.03MB/s    0:00:00 (xfer#1, to-check=0/1)\n",
      "\n",
      "sent 13646 bytes  received 9182 bytes  15218.67 bytes/sec\n",
      "total size is 5133989  speedup is 224.90\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gunzip: raw_data/dmel/transcriptome/dm6.ncbiRefSeq.gtf already exists -- skipping\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='chflags nohidden raw_data/dmel/transcriptome/dm6.ncbiRefSeq.gtf', returncode=0)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_dir = \"input/dmel/transcriptome\"\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "# Download the transcriptome \n",
    "rsync_path = \"rsync://hgdownload.soe.ucsc.edu/goldenPath/dm6/bigZips/genes/dm6.ncbiRefSeq.gtf.gz\"\n",
    "output_path = os.path.join(output_dir, \"dm6.ncbiRefSeq.gtf.gz\")\n",
    "command = f\"rsync --partial --progress {rsync_path} {output_path}\"\n",
    "subprocess.run(command, shell=True)\n",
    "\n",
    "# Unzip the transcriptome \n",
    "output_path = \"input/dmel/transcriptome/dm6.ncbiRefSeq.gtf.gz\"\n",
    "command = f\"gunzip {output_path}\"\n",
    "subprocess.run(command, shell=True)\n",
    "\n",
    "# Unhide transcriptome file\n",
    "output_path = \"input/dmel/transcriptome/dm6.ncbiRefSeq.gtf\"\n",
    "command = f\"chflags nohidden {output_path}\"\n",
    "subprocess.run(command, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Build transcriptome object "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 17868 unique genes.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 17868/17868 [00:23<00:00, 746.26it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcriptome(genes=17868)\n",
      "Transcriptome object has been updated and saved to mel_transcriptome.pkl\n"
     ]
    }
   ],
   "source": [
    "mel_genome_path = \"input/dmel/genome/dm6.fa\"\n",
    "mel_transcriptome_path = \"input/dmel/transcriptome/dm6.ncbiRefSeq.gtf\"\n",
    "mel_object_name = \"mel_transcriptome\"\n",
    "update_transcriptome_object(mel_genome_path, mel_transcriptome_path, mel_object_name, 'dmel')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "tr = load_transcriptome_object(mel_object_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "check_exons_contain_all_features(tr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export mRNA to Fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export mRNA sequences without introns\n",
    "output_dir = \"input/dmel/transcriptome/mRNA_no_introns\"\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "output_path = os.path.join(output_dir, \"mRNA_no_introns.fasta\")\n",
    "\n",
    "with open(output_path, \"w\") as output_file:\n",
    "    for gene_name in tr.genes.keys():\n",
    "        gene = tr.genes[gene_name]\n",
    "        for transcript in gene.transcripts:\n",
    "            output_file.write(f\">{transcript.name} gene={gene.name} location={transcript.chromosome}:{transcript.position[0]}-{transcript.position[1]} strand={transcript.strand} \\n{transcript.mrna_sequence} \\n\")\n",
    "\n",
    "\n",
    "# Export mRNA sequences with introns\n",
    "output_dir = \"input/dmel/transcriptome/mRNA_yes_introns\"\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "output_path = os.path.join(output_dir, \"mRNA_yes_introns.fasta\")\n",
    "\n",
    "with open(output_path, \"w\") as output_file:\n",
    "    for gene_name in tr.genes.keys():\n",
    "        gene = tr.genes[gene_name]\n",
    "        for transcript in gene.transcripts:\n",
    "            output_file.write(f\">{transcript.name} gene={gene.name} location={transcript.chromosome}:{transcript.position[0]}-{transcript.position[1]} strand={transcript.strand} \\n{transcript.dna_sequence} \\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create Blast Databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "makeblastdb: 2.15.0+\n",
      " Package: blast 2.15.0, build Oct 19 2023 15:16:13\n"
     ]
    }
   ],
   "source": [
    "# Check that makeblastdb is installed\n",
    "!makeblastdb -version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 12/02/2024 08:53:02\n",
      "New DB name:   /Users/giacomo.glotzer/Desktop/Drosophila-HCR/raw_data/dmel/transcriptome/mRNA_no_introns/mRNA_no_introns\n",
      "New DB title:  raw_data/dmel/transcriptome/mRNA_no_introns/mRNA_no_introns.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Users/giacomo.glotzer/Desktop/Drosophila-HCR/raw_data/dmel/transcriptome/mRNA_no_introns/mRNA_no_introns\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 35121 sequences in 0.883459 seconds.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 12/02/2024 08:53:04\n",
      "New DB name:   /Users/giacomo.glotzer/Desktop/Drosophila-HCR/raw_data/dmel/transcriptome/mRNA_yes_introns/mRNA_yes_introns\n",
      "New DB title:  raw_data/dmel/transcriptome/mRNA_yes_introns/mRNA_yes_introns.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Users/giacomo.glotzer/Desktop/Drosophila-HCR/raw_data/dmel/transcriptome/mRNA_yes_introns/mRNA_yes_introns\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 35121 sequences in 1.79632 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='makeblastdb -in raw_data/dmel/transcriptome/mRNA_yes_introns/mRNA_yes_introns.fasta -dbtype nucl -parse_seqids -out raw_data/dmel/transcriptome/mRNA_yes_introns/mRNA_yes_introns', returncode=0)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make blast database without introns \n",
    "input_path = \"input/dmel/transcriptome/mRNA_no_introns/mRNA_no_introns.fasta\"\n",
    "output_path = \"input/dmel/transcriptome/mRNA_no_introns/mRNA_no_introns\"\n",
    "command = f\"makeblastdb -in {input_path} -dbtype nucl -parse_seqids -out {output_path}\"\n",
    "subprocess.run(command, shell=True)\n",
    "\n",
    "# Make blast database with introns\n",
    "input_path = \"input/dmel/transcriptome/mRNA_yes_introns/mRNA_yes_introns.fasta\"\n",
    "output_path = \"input/dmel/transcriptome/mRNA_yes_introns/mRNA_yes_introns\"\n",
    "command = f\"makeblastdb -in {input_path} -dbtype nucl -parse_seqids -out {output_path}\"\n",
    "subprocess.run(command, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example: Drosophila Yakuba\n",
    "- https://www.ncbi.nlm.nih.gov/refseq/annotation_euk/Drosophila_yakuba/102/\n",
    "- ASSEMBLY NAME:\tPrin_Dyak_Tai18E2_2.1\n",
    "- ASSEMBLY ACCESSION:\tGCF_016746365.2\n",
    "- ASSEMBLY SUBMITTER:\tPrinceton University\n",
    "- ASSEMBLY DATE:\t30 July 2021\n",
    "- ASSEMBLY TYPE:\tHaploid"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Warning Notice!\n",
      "\n",
      "You are accessing a U.S. Government information system which includes this\n",
      "computer, network, and all attached devices. This system is for\n",
      "Government-authorized use only. Unauthorized use of this system may result in\n",
      "disciplinary action and civil and criminal penalties. System users have no\n",
      "expectation of privacy regarding any communications or data processed by this\n",
      "system. At any time, the government may monitor, record, or seize any\n",
      "communication or data transiting or stored on this information system.\n",
      "\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "Welcome to the NCBI rsync server.\n",
      "\n",
      "\n",
      "GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.fna.gz\n",
      "    45811814 100%   18.69MB/s    0:00:02 (xfer#1, to-check=0/1)\n",
      "\n",
      "sent 38 bytes  received 45823139 bytes  18329270.80 bytes/sec\n",
      "total size is 45811814  speedup is 1.00\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='mv raw_data/dyak/genome/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.fna raw_data/dyak/genome/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.fa', returncode=0)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make directory \n",
    "output_dir = \"input/dyak/genome\"\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "# Download the genome \n",
    "rsync_path = \"rsync://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/016/746/365/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.fna.gz\"\n",
    "output_path = os.path.join(output_dir, \"GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.fna.gz\")\n",
    "command = f\"rsync --partial --progress {rsync_path} {output_path}\"\n",
    "subprocess.run(command, shell=True)\n",
    "\n",
    "\n",
    "# Unzip the genome\n",
    "output_path = \"input/dyak/genome/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.fna.gz\"\n",
    "command = f\"gunzip {output_path}\"\n",
    "subprocess.run(command, shell=True)\n",
    "\n",
    "# Unhide genome file \n",
    "output_path = \"input/dyak/genome/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.fna\"\n",
    "command = f\"chflags nohidden {output_path} {output_path}\" \n",
    "subprocess.run(command, shell=True)\n",
    "\n",
    "# convert from fna to fa \n",
    "input_path = \"input/dyak/genome/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.fna\"\n",
    "output_path = \"input/dyak/genome/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.fa\"\n",
    "command = f\"mv {input_path} {output_path}\"\n",
    "subprocess.run(command, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download Transcriptome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Warning Notice!\n",
      "\n",
      "You are accessing a U.S. Government information system which includes this\n",
      "computer, network, and all attached devices. This system is for\n",
      "Government-authorized use only. Unauthorized use of this system may result in\n",
      "disciplinary action and civil and criminal penalties. System users have no\n",
      "expectation of privacy regarding any communications or data processed by this\n",
      "system. At any time, the government may monitor, record, or seize any\n",
      "communication or data transiting or stored on this information system.\n",
      "\n",
      "-------------------------------------------------------------------------------\n",
      "\n",
      "Welcome to the NCBI rsync server.\n",
      "\n",
      "\n",
      "GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.gtf.gz\n",
      "     5165818 100%  447.86MB/s    0:00:00 (xfer#1, to-check=0/1)\n",
      "\n",
      "sent 13682 bytes  received 9237 bytes  15279.33 bytes/sec\n",
      "total size is 5165818  speedup is 225.39\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gunzip: raw_data/dyak/transcriptome/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.gtf already exists -- skipping\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='chflags nohidden raw_data/dyak/transcriptome/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.gtf', returncode=0)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_dir = \"input/dyak/transcriptome\"\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "# Download the transcriptome \n",
    "rsync_path = \"rsync://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/016/746/365/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.gtf.gz\"\n",
    "output_path = os.path.join(output_dir, \"GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.gtf.gz\")\n",
    "command = f\"rsync --partial --progress {rsync_path} {output_path}\"\n",
    "subprocess.run(command, shell=True)\n",
    "\n",
    "# Unzip the transcriptome \n",
    "output_path = \"input/dyak/transcriptome/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.gtf.gz\"\n",
    "command = f\"gunzip {output_path}\"\n",
    "subprocess.run(command, shell=True)\n",
    "\n",
    "# Unhide transcriptome file\n",
    "output_path = \"input/dyak/transcriptome/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.gtf\"\n",
    "command = f\"chflags nohidden {output_path}\"\n",
    "subprocess.run(command, shell=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Build transcriptome object "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 16150 unique genes.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16150/16150 [00:21<00:00, 762.36it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcriptome(genes=16150)\n",
      "Transcriptome object has been updated and saved to yak_transcriptome.pkl\n"
     ]
    }
   ],
   "source": [
    "yak_genome_path = \"input/dyak/genome/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.fa\"\n",
    "yak_transcriptome_path = \"input/dyak/transcriptome/GCF_016746365.2_Prin_Dyak_Tai18E2_2.1_genomic.gtf\"\n",
    "yak_object_name = \"yak_transcriptome\"\n",
    "update_transcriptome_object(yak_genome_path, yak_transcriptome_path, yak_object_name, 'dyak')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "tr = load_transcriptome_object(yak_object_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "check_exons_contain_all_features(tr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export mRNA to Fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export mRNA without introns\n",
    "output_dir = \"input/dyak/transcriptome/mRNA_no_introns\"\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "output_path = os.path.join(output_dir, \"mRNA_no_introns.fasta\")\n",
    "\n",
    "with open(output_path, \"w\") as output_file:\n",
    "    for gene_name in tr.genes.keys():\n",
    "        gene = tr.genes[gene_name]\n",
    "        for transcript in gene.transcripts:\n",
    "            output_file.write(f\">{transcript.name} gene={gene.name} location={transcript.chromosome}:{transcript.position[0]}-{transcript.position[1]} strand={transcript.strand} \\n{transcript.mrna_sequence} \\n\")\n",
    "\n",
    "\n",
    "# Export mRNA with introns \n",
    "output_dir = \"input/dyak/transcriptome/mRNA_yes_introns\"\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "output_path = os.path.join(output_dir, \"mRNA_yes_introns.fasta\")\n",
    "\n",
    "with open(output_path, \"w\") as output_file:\n",
    "    for gene_name in tr.genes.keys():\n",
    "        gene = tr.genes[gene_name]\n",
    "        for transcript in gene.transcripts:\n",
    "            output_file.write(f\">{transcript.name} gene={gene.name} location={transcript.chromosome}:{transcript.position[0]}-{transcript.position[1]} strand={transcript.strand} \\n{transcript.dna_sequence} \\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create Blast Databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 12/02/2024 08:53:38\n",
      "New DB name:   /Users/giacomo.glotzer/Desktop/Drosophila-HCR/raw_data/dyak/transcriptome/mRNA_no_introns/mRNA_no_introns\n",
      "New DB title:  raw_data/dyak/transcriptome/mRNA_no_introns/mRNA_no_introns.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Users/giacomo.glotzer/Desktop/Drosophila-HCR/raw_data/dyak/transcriptome/mRNA_no_introns/mRNA_no_introns\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 28247 sequences in 0.8102 seconds.\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 12/02/2024 08:53:39\n",
      "New DB name:   /Users/giacomo.glotzer/Desktop/Drosophila-HCR/raw_data/dyak/transcriptome/mRNA_yes_introns/mRNA_yes_introns\n",
      "New DB title:  raw_data/dyak/transcriptome/mRNA_yes_introns/mRNA_yes_introns.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Users/giacomo.glotzer/Desktop/Drosophila-HCR/raw_data/dyak/transcriptome/mRNA_yes_introns/mRNA_yes_introns\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 28247 sequences in 1.81686 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='makeblastdb -in raw_data/dyak/transcriptome/mRNA_yes_introns/mRNA_yes_introns.fasta -dbtype nucl -parse_seqids -out raw_data/dyak/transcriptome/mRNA_yes_introns/mRNA_yes_introns', returncode=0)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_path = \"input/dyak/transcriptome/mRNA_no_introns/mRNA_no_introns.fasta\"\n",
    "output_path = \"input/dyak/transcriptome/mRNA_no_introns/mRNA_no_introns\"\n",
    "command = f\"makeblastdb -in {input_path} -dbtype nucl -parse_seqids -out {output_path}\"\n",
    "subprocess.run(command, shell=True)\n",
    "\n",
    "input_path = \"input/dyak/transcriptome/mRNA_yes_introns/mRNA_yes_introns.fasta\"\n",
    "output_path = \"input/dyak/transcriptome/mRNA_yes_introns/mRNA_yes_introns\"\n",
    "command = f\"makeblastdb -in {input_path} -dbtype nucl -parse_seqids -out {output_path}\"\n",
    "subprocess.run(command, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get Chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Number of Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_052530.2</td>\n",
       "      <td>3822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_052528.2</td>\n",
       "      <td>3212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_052527.2</td>\n",
       "      <td>3192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_052529.2</td>\n",
       "      <td>3068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_052526.2</td>\n",
       "      <td>2368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_052531.2</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NW_025048787.1</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NW_025048801.1</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NW_025048817.1</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_001322.1</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  Number of Genes\n",
       "0     NC_052530.2             3822\n",
       "1     NC_052528.2             3212\n",
       "2     NC_052527.2             3192\n",
       "3     NC_052529.2             3068\n",
       "4     NC_052526.2             2368\n",
       "5     NC_052531.2              107\n",
       "6  NW_025048787.1               63\n",
       "7  NW_025048801.1               42\n",
       "8  NW_025048817.1               40\n",
       "9     NC_001322.1               37"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the chromosomes and number of genes per chromosome \n",
    "chromosomes = {}\n",
    "for gene_name in tr.genes.keys():\n",
    "    gene = tr.get_gene(gene_name)\n",
    "    chromosomes[gene.chromosome] = chromosomes.get(gene.chromosome, 0) + 1\n",
    "chrom_counts = pd.DataFrame(chromosomes.items(), columns=['Chromosome', 'Number of Genes'])\n",
    "chrom_counts = chrom_counts.sort_values(by='Number of Genes', ascending=False)\n",
    "chrom_counts = chrom_counts.reset_index(drop=True)\n",
    "chrom_counts[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hcr-probe-design-general-py3.11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
